Extractors -> listeners, trigger listeners on saved feeds automatically

max-zilla opened the pull request:
This renames extractors to listeners (and creates a legacy endpoint for old extractors) as well as includes basic implementation of job feeds (saved searches) and listeners on those feeds. Lots of decisions here we should discuss.

See test_feeds.py for implementation example, still in progress and very rough.

tcnichol reviewed (approved):
Ran this one and it still works with extractors. Looks good. Marking approved.

tcnichol commented:
I think the failing test might be due to ListenerIn needing to be a superclass of ExtractorInfo instead of BaseModel, or perhaps the other way around.

lmarini commented:
Will you be adding https://github.com/clowder-framework/clowder2/blob/workflow-documentation/backend/docs/source/listeners.md to this PR?

lmarini commented:
Are you still in favor or renaming `Listeners` to `EventListeners` to be a bit more specific. When looking through the code and documentation I am still wondering if `Listeners` is still too generic, even for a developer.

max-zilla commented:
> Can you add docstrings to all functions and models? We need to start requiring this on all PRs. Thanks!

Added a bunch of these.

> Will you be adding https://github.com/clowder-framework/clowder2/blob/workflow-documentation/backend/docs/source/listeners.md to this PR?

Added with minor updates as well.

> Are you still in favor or renaming `Listeners` to `EventListeners` to be a bit more specific. When looking through the code and documentation I am still wondering if `Listeners` is still too generic, even for a developer.

I renamed to EventListeners across the app, although "listeners.py" is still the filename for simplicity.

lmarini commented on backend/app/models/listeners.py:
```diff
@@ -0,0 +1,88 @@
+    parameters: List[dict] = []
+
+
+class EventListenerBase(BaseModel):
```
Is this supposed to inherit from `ExtractorInfo`?

lmarini replied:
Sounds good. We can try to keep both. Maybe add a few more comments explaining the difference? As long as we can submit/trigger with current extractors, we are good short term. I don't think this gets in the way of that?

lmarini commented on backend/app/models/listeners.py:
```diff
@@ -0,0 +1,88 @@
+    """v1 Extractors can submit data formatted as a LegacyEventListener (i.e. v1 format) and it will be converted to a v2 EventListener."""
+
+    name: str
+    version: str = "1.0"
```
Why the difference in types between version for `EventListenerBase` and `LegacyEventListener` (str vs int)?

max-zilla replied:
Clowder v1 lets users arbitrarily provide a version with no rules (e.g. "1", "1.0", "v1" "beta" all allowed). I was thinking (before v1 back-compatibility) that version could just be an integer for clarity and simplicity, but after seeing it is just a string field for v1 we should probably use that. 

If we use string, "version" might better be called "tag" or something like Docker does? Don't necessarily see a point in enforcing something that can be cast as a float in that case...

lmarini replied:
> Clowder v1 lets users arbitrarily provide a version with no rules (e.g. "1", "1.0", "v1" "beta" all allowed). I was thinking (before v1 back-compatibility) that version could just be an integer for clarity and simplicity, but after seeing it is just a string field for v1 we should probably use that.
> 
> If we use string, "version" might better be called "tag" or something like Docker does? Don't necessarily see a point in enforcing something that can be cast as a float in that case...

Good point. Tag sounds great. We could also enforce semantic versioning in the string and call it version. But that might be more work. Similar to what npm does.

lmarini commented on backend/app/tests/test_feeds.py:
```diff
@@ -0,0 +1,70 @@
+    os.remove(dummy_file)
+    assert response.status_code == 200
+
+    # Verify the message
```
Is there a way to test that the listener was triggered?

max-zilla replied:
not yet. if we had an event log we could check that but it doesn't exist yet. Otherwise we'd have to check RMQ somehow.

lmarini replied:
@tcnichol how easily could you add v1 events back into v2? I am looking at the heartbeat PR and will leave comments there soon. It would be a similar approach?

lmarini commented on backend/app/routers/feeds.py:
```diff
@@ -0,0 +1,158 @@
+
+router = APIRouter()
+
+clowder_bucket = os.getenv("MINIO_BUCKET_NAME", "clowder")
```
I don't think this is being used.

lmarini commented on backend/app/routers/feeds.py:
```diff
@@ -0,0 +1,158 @@
+                        listeners_found.append(listener.listener_id)
+
+    for targ_listener in listeners_found:
+        queue = ""  # TODO: Each extractor gets a queue - routing key same as name?
```
Can this default to the listener name so that legacy extractors can be triggered?

lmarini reviewed (changes requested):
Thanks for updating the PR. Left a few more questions / comments as I continue to review this. There is a lot of digest. Thanks!
